### PR TITLE
Introduce cache passive state

### PR DESCRIPTION
**Pull request by robertbaldyga**

### CODE REVIEW

File: src/mngt/ocf_mngt_cache.c [hunk 1]
```diff
@@ -1562,6 +1583,79 @@ struct ocf_pipeline_properties _ocf_mngt_cache_load_pipeline_properties = {
...
+	.finish = _ocf_mngt_cache_attach_finish,
+	.steps = {
+		OCF_PL_STEP(_ocf_mngt_copy_uuid_data),
+		OCF_PL_STEP(_ocf_mngt_init_attached_nonpersistent),
+		OCF_PL_STEP(_ocf_mngt_attach_cache_device),
+		OCF_PL_STEP(_ocf_mngt_init_cache_front_volume),
+		OCF_PL_STEP(_ocf_mngt_init_properties),
+		OCF_PL_STEP(_ocf_mngt_attach_check_ram),
+		OCF_PL_STEP(_ocf_mngt_test_volume),
+		OCF_PL_STEP(_ocf_mngt_init_properties),
```
Comment (mmichal10) [inline src/mngt/ocf_mngt_cache.c:1653]: Is `_ocf_mngt_init_properties()` called twice on purpose?

File: src/ocf_cache.c [hunk 2]
```diff
@@ -231,3 +231,197 @@ void *ocf_cache_get_priv(ocf_cache_t cache)
...
+
+
+
+
+
+
+
+
+
+
```
Comment (mmichal10) [inline src/ocf_cache.c:248]: Redundant newlines

File: src/mngt/ocf_mngt_cache.c (hunk 1 above)
Comment (robertbaldyga) [reply src/mngt/ocf_mngt_cache.c:1655]: No. Fixed.

File: src/ocf_cache.c (hunk 2 above)
Comment (robertbaldyga) [reply src/ocf_cache.c:248]: Fixed.

File: src/mngt/ocf_mngt_cache.c [hunk 3]
```diff
@@ -1921,6 +1933,168 @@ struct ocf_pipeline_properties _ocf_mngt_cache_bind_pipeline_properties = {
...
+static void _ocf_mngt_activate_swap_stop_pipeline(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_cache_attach_context *context = priv;
+	ocf_cache_t cache = context->cache;
+	ocf_pipeline_t stop_pipeline;
+	int result;
+
+	result = ocf_pipeline_create(&stop_pipeline, cache,
+			&ocf_mngt_cache_stop_pipeline_properties);
```
Comment (mmichal10) [inline src/mngt/ocf_mngt_cache.c:2033]: If `ocf_pipeline_create()` fails, `ocf_mngt_cache_stop_passive_pipeline_properties`  isn't suitable to stop the cache at this point; i.e. it doesn't handle removing cores thus stop pipelines have to swapped earlier. Also the stop pipeline has be created in `_ocf_mngt_cache_activate()`

File: src/mngt/ocf_mngt_cache.c [hunk 4]
```diff
@@ -2327,6 +2530,51 @@ void ocf_mngt_cache_bind(ocf_cache_t cache,
...
+static void _ocf_mngt_cache_activate_complete(ocf_cache_t cache, void *priv1,
+		void *priv2, int error)
+{
+	ocf_mngt_cache_bind_end_t cmpl = priv1;
+
+	if (error)
+		OCF_CMPL_RET(cache, priv2, error);
+
+	_ocf_mngt_cache_set_active(cache);
+	ocf_cache_log(cache, log_info, "Successfully binded\n");
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply src/mngt/ocf_mngt_cache.c:2542]: Oops, copy-paste not smart enough yet.

File: src/mngt/ocf_mngt_cache.c (hunk 3 above)
Comment (robertbaldyga) [reply src/mngt/ocf_mngt_cache.c:2033]: This is not a problem as `_ocf_mngt_cache_activate_finish()` handles cleanup after pipeline failure, so we leave in the same state as we started (i.e. _passive_ state).

File: src/mngt/ocf_mngt_cache.c [hunk 5]
```diff
@@ -362,8 +362,21 @@ static void _ocf_mngt_load_add_cores(ocf_pipeline_t pipeline,
...
 		ocf_volume_t tvolume = NULL;
 
+		muuid = ocf_metadata_get_core_uuid(cache, core_id);
+		uuid.data = muuid->data;
+		uuid.size = muuid->size;
+
+		volume_type = ocf_ctx_get_volume_type(cache->owner,
+				core->conf_meta->type);
+
+		ocf_volume_init(&core->volume, volume_type, &uuid, false);
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply src/mngt/ocf_mngt_cache.c:385]: I agree, we can simplify this. Let's do one thing at the time though (which in this case is moving volume initialization from metadata code to cache mngt).

**Comment by mmichal10** [comment]
As far as I remember you were supposed to extract volume related changes into separate patchset

**Review by mmichal10 — APPROVED** [review]

**Review by mmichal10 — APPROVED** [review]

**Review by jfckm — APPROVED** [review]

**Pull request merged**